### PR TITLE
watson ML ibm_watsonx_ai package support  and multiple response support

### PR DESCRIPTION
**Pull request by huang-cn**
Fixed https://github.com/traceloop/openllmetry/issues/441

### CODE REVIEW

**Comment by nirga** [comment]
Thanks @huang-cn! As the number of contributors is increasing, I'd suggest if you can add a couple of tests for that, similar to other instrumentations. Otherwise this may break it some point and we won't know about it.

**Comment by huang-cn** [comment]
> Thanks @huang-cn! As the number of contributors is increasing, I'd suggest if you can add a couple of tests for that, similar to other instrumentations. Otherwise this may break it some point and we won't know about it.

@nirga Sure, I'll add some tests.

**Comment by nirga** [comment]
> There's some check failed and I saw message in "CI / Build & Test Packages (3.8) (pull_request)":
> 
> ```
> tests/conftest.py:5: in <module>
>     from opentelemetry.sdk.trace import TracerProvider
> E   ModuleNotFoundError: No module named 'opentelemetry.sdk'
> ```
> 
> not sure why it happened, @nirga could you let us know what we can do to solve this?

@huang-cn I think you're missing [some dependencies](https://github.com/traceloop/openllmetry/blob/cd3384f7f034b4b679306205cb15f2dc1fbc9a54/packages/opentelemetry-instrumentation-openai/pyproject.toml#L34)

**Comment by huang-cn** [comment]
@nirga Thanks! I'll fix the dependencies.

File: packages/opentelemetry-instrumentation-watsonx/tests/cassettes/test_generate.yaml [hunk 1]
```diff
@@ -0,0 +1,709 @@
+interactions:
+- request:
+    body: apikey=test_api_key&grant_type=urn%3Aibm%3Aparams%3Aoauth%3Agrant-type%3Aapikey
```
[the remark this answers is not in the document]
Comment (huang-cn) [reply packages/opentelemetry-instrumentation-watsonx/tests/cassettes/test_generate.yaml]: @nirga I was using `pytest-vcr`, it seems incompatible with `pytest-recording`.  now I switched to `pytest-recording` and recorded new cassettes, I think it should works this time.

**Comment by nirga** [comment]
Tests seem to be flaky 🤔

**Comment by huang-cn** [comment]
I've checked, it's likely because the token expired, I'm working on it.

**Comment by huang-cn** [comment]
@nirga the test can run locally but not in CI, strange. I switched to use a different approach, changing the access_token.

File: packages/opentelemetry-instrumentation-watsonx/tests/cassettes/test_generate/test_generate.yaml [hunk 2]
```diff
@@ -0,0 +1,615 @@
...
...so_8FWS2Wv29cjMrIEe2fMZBAGP4d6BYicITrExJoOblWwRyIq-ahPSYWwJ-dJI6lbYXefBtK2Om-7a8NtuQ8a1qRlL5NHmD8Ks42GxuRzRiGv8lV_OdySgBSm6cLLgPquLTU8ERreKnZeV6b9e8P-UiYEaWgz1HvvWAproZFu9e0zCWwPQU9dR2ZpecEssX9t0O9Eo1lnittvv_LpXR5yOISMKuYfT8ipawyOxNtGE7fuh05sBrzsmj0Uw1auy36uojzMKO4oSqkDI6mTFn_z0VBOIY1Zh3ly0g9VQEQlFwt-sWK2MshFD2Hjez7-AWT_ydorFL2497TGBn5JpiNcdExlZYXuRGXl4HY5c-sWbdVNb14Yhft1h6Fof0YhASnwH0k3JCQfo_ec_P6NMZeypK-fZLe5ma4H3Z1lo3Jji7J0zMwbPpWjcUheUlgLqJb2Pyr3k9GQxu9bErnHShIUdxM-4x3JNnKe-PwwyxMW9VR_cglwnYfWqflXBtttnmYGgSKc5KEZMuY6-Jj6HhwDF4RUlc6mFK_v_1KVtQUyULZbGX6ctZKBxSysgl6p7mbDbzdLwLNZisulITyCt-LQzWsLL4WWOQxRQF-UykGR63_6temZ--EJi9wRmm9fJ5qdfCHdsI9v1pA3LOuNFsPfJE63TeYCdG2kf7JF_2ElLo_IaQTDQ5TunU4GV-LLoc6b-7dhFdI8fQfGvO4ZPWLFQZ9lCfDUiGmbkEWUhgy59lWtKVuqpVB2qJWKYwUg-P41m_zFlIETj4sz0ELNtR32H6F9I5A6F5vvX2P30wAvW2PBsSnWE474fb5EUI_ejweOyEz-067Qis4JIUsguIbmI8-JDuMDobSbyFA2_ujkUBjcbRqrQs4bY86hbu-w_-Wqf1qbKYxpV8Mb3SQLM0PaCAhCjbq9o5gcfhUhP-9HWgCla04BVTPqrf7nJc270bcyKb_HDar3SiTYgxDe-WFCXP2uX_QS1EziqkP-uPvnlaz1yjC8Uxw76MzuTbmPxllcH8RuOnAz2ai-9nXKM2RiN06D62bE0mc-NB1k1Q8ZYr6YsW-rAPIJ6BC8NL","token_type":"Bearer","expires_in":3600,"expiration":1708597208,"refresh_token_expiration":1708852808,"scope":"ibm
```
Comment (paolorechia) [inline packages/opentelemetry-instrumentation-watsonx/tests/cassettes/test_generate/test_generate.yaml:21]: ❓ Are you sure it’s OK (from your side) committing these credentials here?

I’m a bit short on time to check carefully, maybe you’re missing a VCR filter for this? (Hence the failure in CI/CD)?

Comment (huang-cn) [reply packages/opentelemetry-instrumentation-watsonx/tests/cassettes/test_generate/test_generate.yaml:21]: @paolorechia this is actually a test jwt token I generated, not the actual one, and there're other api keys in this yaml, but there were long expired (within an hour), this yaml was recorded some hours ago, so no need to worry about.

Comment (huang-cn) [reply packages/opentelemetry-instrumentation-watsonx/tests/cassettes/test_generate/test_generate.yaml:21]: those failures were caused by a token self validation, b/c the jwt token encoded with an expiration time(an hour), so I workaround it by creating a test token that never expire.

Comment (paolorechia) [reply packages/opentelemetry-instrumentation-watsonx/tests/cassettes/test_generate/test_generate.yaml:21]: Nice find! I saw now they passed

**Comment by huang-cn** [comment]
> Could you share some E2E test results? Screenshots or logged traces?

Sure.  
Here is the screenshot of langchain trace:  
image

traceloop flow:  
image

File: packages/opentelemetry-instrumentation-watsonx/opentelemetry/instrumentation/watsonx/__init__.py [hunk 3]
```diff
@@ -77,68 +97,132 @@ def should_send_prompts():
...
+        )
+        _set_span_attribute(
+            span,
+            f"{SpanAttributes.LLM_USAGE_PROMPT_TOKENS}.{index}",
+            usage.get("input_token_count"),
+        )
+        _set_span_attribute(
+            span,
+            f"{SpanAttributes.LLM_COMPLETIONS}.{index}.content",
+            response["results"][0]["generated_text"],
```
[the remark this answers is not in the document]
Comment (huang-cn) [reply packages/opentelemetry-instrumentation-watsonx/opentelemetry/instrumentation/watsonx/__init__.py:185]: Yes, there are response length check before that, the index is enumerate index,  it should be safe to use.

File: packages/opentelemetry-instrumentation-watsonx/opentelemetry/instrumentation/watsonx/__init__.py [hunk 4]
```diff
@@ -31,23 +31,43 @@
...
-        "module": "ibm_watson_machine_learning.foundation_models",
-        "object": "Model",
+        "module": "ibm_watsonx_ai.foundation_models",
+        "object": "ModelInference",
         "method": "get_details",
         "span_name": "watsonx.get_details",
     },
 ]
 
+WATSON_MODUELS = [
```
[the remark this answers is not in the document]
Comment (huang-cn) [reply packages/opentelemetry-instrumentation-watsonx/opentelemetry/instrumentation/watsonx/__init__.py]: I've fixed that.

File: packages/opentelemetry-instrumentation-watsonx/tests/test_generate.py [hunk 5]
```diff
@@ -0,0 +1,18 @@
+import pytest
+
+
+@pytest.mark.vcr
```
[the remark this answers is not in the document]
Comment (huang-cn) [reply packages/opentelemetry-instrumentation-watsonx/tests/test_generate.py:4]: Yes, there will be more tests to be added here, in separated PR probably.

**Review by paolorechia — APPROVED** [review]
Thanks for the work! LGTM

**Pull request merged**